Homework #3. Task #3. Волынец Арсений

vol0n opened the pull request.

vol0n committed:
Initial commit of hw3_task3

cupertank commented on src/main/kotlin/homework3/task_3.kt:
```diff
@@ -0,0 +1,119 @@
+    @SerialName("functions")
+    val functions: MutableList<Function>?
+) {
+    init {
+        if (removeDoubles(functions)) {
+            println("Warning: yaml config contains identical functions. Only one function is generated.")
+        }
+    }
```
Мне кажется, что можно просто взять Set от functions, записать его в классе, сравнить размер сета и исходного массива и вывести Warning. Просто твоя функция по удалению дубликатов выглядит как-то очень дико, да и назвать её стоило бы removeDuplicates

cupertank commented on src/main/kotlin/homework3/task_3.kt:
```diff
@@ -0,0 +1,119 @@
+        val config = Yaml.default.decodeFromString(YamlConfig.serializer(), configString)
+        val name = ClassName(config.packageName, config.className + "Test")
+        val type = TypeSpec.classBuilder(name)
+        if (config.functions !== null) {
```
Можно просто !=, так ты проверяешь ссылку, так не пишут просто

cupertank commented on src/test/kotlin/homework3/Task_3KtTest.kt:
```diff
@@ -0,0 +1,31 @@
+    @MethodSource("inputData")
+    @ParameterizedTest(name = "test {index}: {0}")
+    fun generateTestFileTest(testConfigName: String, expectedName: String, packageName: String) {
+        val pathToResources = "./src/test/resources/homework3"
```
Такое никуда не годится, почитай про работу с ресурсами в Java/Kotlin'е

cupertank commented on src/main/kotlin/homework3/task_3.kt:
```diff
@@ -0,0 +1,119 @@
+        }
+        val file = FileSpec.builder(config.packageName, config.className)
+            .addType(type.build()).build()
+        File(locationToSave).apply {
+                file.writeTo(this)
```
Кажется у вас в примере строятся internal классы, а у тебя public, доработай

cupertank commented on src/test/kotlin/homework3/Task_3KtTest.kt:
```diff
@@ -0,0 +1,31 @@
+class Task_3KtTest {
+    companion object{
+        @JvmStatic
+        fun inputData(): List<Arguments> = listOf(
+            Arguments.of("noFunctions.yaml", "noFunctions.txt", "noFunctions"),
+            Arguments.of("identicalFuncs.yaml", "identicalFuncs.txt", "identicalFuncs"),
+            Arguments.of("yamlTest.yaml", "yamlTest.txt", "yamlTest")
```
Я предлагаю унифицировать все немного иначе. Сделай в ресурсах для каждого теста отдельную папку с говорящим названием, в каждой папке сделай config.yaml и expected.kt(txt). По-итогу в аргументы параметризованных тестов ты просто передашь название папок, а по-совместительству и название теста))

cupertank commented on src/test/kotlin/homework3/Task_3KtTest.kt:
```diff
@@ -0,0 +1,31 @@
+    @ParameterizedTest(name = "test {index}: {0}")
+    fun generateTestFileTest(testConfigName: String, expectedName: String, packageName: String) {
+        val pathToResources = "./src/test/resources/homework3"
+        generateTestFile("$pathToResources/yamlConfigs/$testConfigName", "$pathToResources/GeneratedTests")
```
Не стоит писать в ресурсы, почитай про Temporary Directory в JUnit 5

cupertank reviewed (changes requested):
Очень неплохо, так держать!!!

cupertank commented:
Подмерджи ещё мастер, чтобы у тебя ещё тесты на винде были. Скорее всего ты увидишь как у тебя они упадут из-за хардкода пути))

cupertank commented on build.gradle.kts:
```diff
@@ -15,14 +16,18 @@ repositories {
 }
 
 dependencies {
+    implementation("org.junit.jupiter:junit-jupiter:5.4.2")
```
Это должно быть testImplementation и у тебя не хватает junit-engine

vol0n replied:
у меня в homework3/task_3.kt есть импорт теста для генерации аннотации к функциям, если тут сделать testImplemention, то homework3/task_3.kt не будет компилиться

vol0n replied:
и почему не хватает junit-engine, он же теперь автоматически подключается, как я понял: https://stackoverflow.com/a/55084036

cupertank reviewed (changes requested):
У тебя абсолютно не запускаются тесты если что, у тебя пустая выдача, потому что ты не до конца настроил грэдл

vol0n committed:
Remove fun removeDoubles, use Set

vol0n committed:
Fix: tests are now running when building

vol0n committed:
Use tempdir, load using getResource

1. For generating tests in Task_3KtTest use junit tempdir
2. Use getResource for accessing yaml configs and expected results

vol0n committed:
Merge branch 'master' into hw3_task3

vol0n committed:
Add step for uploading test report

vol0n committed:
Add always for upload step

vol0n committed:
Artifact dir: build/reports/tests/test/index.html

vol0n committed:
Pass config as string

vol0n committed:
Fix path for uploading artifact: upload all files

vol0n commented:
Не могу понять, почему тесты на винде падают, вроде же текст совпадает.

vol0n committed:
Add logs in tests

vol0n committed:
Try .kt expected files

vol0n committed:
Try: tell git not to modify .txt files

vol0n committed:
Fix wrong index

cupertank commented on src/main/kotlin/homework3/task_3.kt:
```diff
@@ -0,0 +1,103 @@
+    @SerialName("functions")
+    val functions: MutableList<Function>?
+) {
+    var setOfFuncs: Set<Function> = emptySet()
```
Зачем тебе тут var ещё с дополнительным геттером. Сделай просто так: `val setOfFucns: Set<Function>`, а в init сделай ветку else где `setOfFuncs = emptySet()`

cupertank commented on src/test/kotlin/homework3/Task_3KtTest.kt:
```diff
@@ -0,0 +1,55 @@
+        println(actual == expected)
+        for (i in 0..min(actual.lastIndex, expected.lastIndex))
+            if (actual[i] != expected[i])
+                println("Chars are not equal: expected: ${expected[i]}, actual: ${actual[i]}")
```
Мне кажется, что это отладочная информация, так что убери её из тестов

cupertank commented on src/test/resources/homework3/yamlConfigs/identicalFuncs.yaml:
```diff
@@ -0,0 +1,5 @@
+package name: identicalFuncs
+class name: IdenticalFuncs
+functions:
+  - name: "backwardApply"
+  - name: "backwardApply"
```
Мне кажется или всё что в папке yamlConfigs не используется в тестах?

cupertank replied:
Если так, то убери их оттуда

cupertank reviewed (changes requested):
Готов поставить тебе полный балл если ты быстро поправишь всё

cupertank commented:
Разрезолви конфликты ещё

vol0n committed:
Fix style, remove debug inf

vol0n committed:
Fix conflicts
